modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/consistency/GridNearReadRepairAbstractFuture.java

```diff
@@ -145,101 +150,94 @@ protected GridNearReadRepairAbstractFuture(
-                    mappings.computeIfAbsent(node, k -> new HashSet<>()).add(key);
-            }
-
-            primaries = primaryNodes;
-
-            for (Map.Entry<ClusterNode, Collection<KeyCacheObject>> mapping : mappings.entrySet()) {
-                ClusterNode node = mapping.getKey();
-
-                GridPartitionedGetFuture<KeyCacheObject, EntryGetResult> fut =
-                    new GridPartitionedGetFuture<>(
-                        ctx,
-                        mapping.getValue(), // Keys.
-                        readThrough,
-                        false, // Local get required.
-                        taskName,
-                        deserializeBinary,
-                        recovery,
-                        expiryPlc,
-                        false,
-                        true,
-                        true,
-                        tx != null ? tx.label() : null,
-                        tx != null ? tx.mvccSnapshot() : null,
-                        node);
-
-                fut.listen(this::onResult);
-
-                futs.put(mapping.getKey(), fut);
-            }
-
             for (GridPartitionedGetFuture<KeyCacheObject, EntryGetResult> fut : futs.values())
                 fut.init(topVer);
-
-            if (futs.isEmpty())
-                onDone(new ClusterTopologyServerNotFoundException("Failed to map keys for cache " +
-                    "(all partition nodes left the grid) [topVer=" + topVer + ", cache=" + ctx.name() + ']'));
         }
         finally {
             ctx.tm().tx(prevTx);
         }
     }
 
     /**
-     * @param topVer Topology version.
+     * @param fut Future to be notified.
      */
-    protected final void remap(AffinityTopologyVersion topVer) {
-        futs.clear();
+    protected final void initOnRemap(GridNearReadRepairAbstractFuture fut) {
+        remapCnt = fut.remapCnt + 1;
+
+        listen(f -> {
+            assert !fut.isDone();
 
-        this.topVer = topVer;
+            fut.onDone(f.result(), f.error());
+        });
 
-        map();
+        init();
     }
 
+    /**
+     * @param topVer Topology version.
+     */
+    protected abstract void remap(AffinityTopologyVersion topVer);
+
     /**
      * Collects results of each 'get' future and prepares an overall result of the operation.
      *
      * @param finished Future represents a result of GET operation.
      */
-    protected final synchronized void onResult(IgniteInternalFuture<Map<KeyCacheObject, EntryGetResult>> finished) {
-        if (isDone() // All subfutures (including currently processing) were successfully finished at previous future processing.
-            || (topVer == null) // Remapping, ignoring any updates until remapped.
-            || !futs.containsValue((GridPartitionedGetFuture<KeyCacheObject, EntryGetResult>)finished)) // Remapped.
-            return;
-
+    protected final void onResult(IgniteInternalFuture<Map<KeyCacheObject, EntryGetResult>> finished) {
```

Comment:
Seems, currently possible to remap the future on each failure.
Future should be remapped only once.